tachyon/math/base/semigroups.h

```diff
@@ -147,12 +157,110 @@ class MultiplicativeSemigroup {
     }
     return g;
   }
+
+  constexpr static std::vector<G> GetSuccessivePowers(size_t size,
+                                                      const G& generator) {
+#if defined(TACHYON_HAS_OPENMP)
+    uint32_t log_size = size_t{base::bits::SafeLog2Ceiling(size)};
+    if (log_size <= kMinLogSizeForParallelization)
+#endif
+      return GetSuccessivePowersSerial(size, generator);
+#if defined(TACHYON_HAS_OPENMP)
+    G power = generator;
+    // [g, g², g⁴, g⁸, ..., g^(2^(|log_size|))]
+    std::vector<G> log_powers = base::CreateVector(log_size, [&power]() {
+      G old_value = power;
+      power.SquareInPlace();
+      return old_value;
+    });
+
+    // allocate the return array and start the recursion
+    std::vector<G> powers =
+        base::CreateVector(size_t{1} << log_size, G::Zero());
+    GetSuccessivePowersRecursive(powers, absl::MakeConstSpan(log_powers));
+    return powers;
+#endif
+  }
+
+ private:
+#if defined(TACHYON_HAS_OPENMP)
+  constexpr static void GetSuccessivePowersRecursive(
+      std::vector<G>& out, const absl::Span<const G>& log_powers) {
+    CHECK_EQ(out.size(), size_t{1} << log_powers.size());
+    // base case: just compute the powers sequentially,
+    // g = log_powers[0], |out| = [1, g, g², ..., g^(|log_powers.size() - 1|)]
+    if (log_powers.size() <= size_t{kMinLogSizeForParallelization}) {
+      out[0] = G::One();
+      for (size_t i = 1; i < out.size(); ++i) {
+        out[i] = out[i - 1] * log_powers[0];
+      }
+      return;
+    }
+
+    // recursive case:
+    // 1. split |log_powers| in half
+    // |log_powers| =[g, g², g⁴, g⁸]
```

Comment:
I don't get why log_powers have inital values of [g, g², g⁴, g⁸]. Is it correct?

Reply:
The recursion gets `[g, g², g⁴, g⁸, ..., g^(2^(|log_size|))]` from the `GetSuccessivePowers`. Then it divides the initial vector into halves and starts Divide and Conquer recursion.
`[g, g², g⁴, g⁸]` In the comment is just a simple example.

Reply:
@Insun35 So what is the |log_size| value for the example case?

Reply:
|log_size| is 3 in this case

Reply:
@Insun35 But like below [here](https://github.com/kroma-network/tachyon/pull/115/files#diff-6b9b9ccc01d7e924806c8c126369e6ace4cccf41ca72f4254a58f49bc17da769R231), it doesn't seem log_size = 3. Can you fix it through all comments?

Reply:
Sorry for the confusion. The |log_size| is 4 in this case. I added some comments to avoid confusion.